src/java.base/share/classes/java/util/stream/Gatherers.java

```diff
@@ -0,0 +1,657 @@
+     *
+     * <p>This operation always emits a single resulting element.
+     *
+     * <p>Example:
+     * {@snippet lang = java:
+     * // will contain: Optional[123456789]
+     * Optional<String> numberString =
+     *     Stream.of(1,2,3,4,5,6,7,8,9)
+     *           .gather(
+     *               Gatherers.fold(() -> "", (string, number) -> string + number)
+     *            )
+     *           .findFirst();
+     * }
+     *
+     * @see java.util.stream.Stream#reduce(Object, BinaryOperator)
+     *
+     * @param initial the identity value for the fold operation
+     * @param folder the folding function
+     * @param <T> the type of elements the returned gatherer consumes
+     * @param <R> the type of elements the returned gatherer produces
+     * @return a new Gatherer
+     * @throws NullPointerException if any of the parameters are null
+     */
+    public static <T, R> Gatherer<T, ?, R> fold(
+            Supplier<R> initial,
+            BiFunction<? super R, ? super T, ? extends R> folder) {
+        Objects.requireNonNull(initial, "'initial' must not be null");
+        Objects.requireNonNull(folder, "'folder' must not be null");
+
+        class State {
+            R value = initial.get();
+            State() {}
+        }
+
+        return Gatherer.ofSequential(
+                State::new,
+                Integrator.ofGreedy((state, element, downstream) -> {
+                    state.value = folder.apply(state.value, element);
+                    return true;
+                }),
+                (state, downstream) -> downstream.push(state.value)
+        );
+    }
+
+    /**
+     * Performs a prefix scan -- an incremental accumulation, using the
+     * provided functions.
+     *
+     * @param initial the supplier of the initial value for the scanner
+     * @param scanner the function to apply for each element
+     * @param <T> the type of element which this gatherer consumes
+     * @param <R> the type of element which this gatherer produces
+     * @return a new Gatherer which performs a prefix scan
+     * @throws NullPointerException if any of the parameters are null
+     */
+    public static <T, R> Gatherer<T, ?, R> scan(
+            Supplier<R> initial,
+            BiFunction<? super R, ? super T, ? extends R> scanner) {
+        Objects.requireNonNull(initial, "'initial' must not be null");
+        Objects.requireNonNull(scanner, "'scanner' must not be null");
+
+        class State {
+            R current = initial.get();
+            boolean integrate(T element, Downstream<? super R> downstream) {
+                return downstream.push(current = scanner.apply(current, element));
+            }
+        }
+
+        return Gatherer.ofSequential(State::new,
+                                     Integrator.<State,T, R>ofGreedy(State::integrate));
+    }
+
+    /**
+     * An operation which executes operations concurrently
+     * with a fixed window of max concurrency, using VirtualThreads.
```

Comment:
Of the 5, this is the one that I'm not sure about so will be interesting to see if there is feedback from real world usage.

"VirtualThreads" isn't a class so maybe replace it with "virtual Threads", linked to Thread.html#virtual-threads.

Reply:
Sounds good. Wilco